### PR TITLE
Support not nested subquery (from SetAssignment and Expression) on UpdateStatement

### LINKED ISSUE #12825: Support not nest subquery on encrypt rewrite
@tristaZero  @strongduanmu 
Support subquery(not nested) on encrypt rewrite, the following sqls was tested (on mysql) ok,  more sqls are welcome to leave a message.

subquery on projection:

logicSQL:
SELECT u.mobile, u.username, u.id_no, (select o.card_no from `order` o where o.card_no=u.id_no) card_no FROM `user` u, `bankcard` c where u.id_no =c.card_no and u.mobile=? and u.id=1

actualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no, (select o.card_no_cipher from `order` o where o.card_no_hash=u.id_no_hash) card_no FROM `user` u, `bankcard` c where u.id_no_hash =c.card_no and u.mobile_hash=? and u.id=1

subquery on tablesegment:

LogicSQL:
SELECT u.mobile, u.username, u.id_no, o.card_no FROM (select card_no from `order`) o, user `u`, bankcard c where o.card_no=u.id_no and o.card_no=c.card_no and u.mobile=? and o.card_no=? and u.id=1

ActualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no, o.card_no_cipher AS card_no FROM (select card_no_cipher, card_no_hash from `order`) o, user `u`, bankcard c where o.card_no_hash=u.id_no_hash and o.card_no_hash=c.card_no and u.mobile_hash=? and o.card_no_hash=? and u.id=1

subquery on expression:

LogicSQL:
SELECT u.mobile, u.username, u.id_no FROM user `u`, bankcard c where u.id_no=(select card_no from `order` where card_no=?) and u.mobile=? and u.id=1

ActualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no FROM user `u`, bankcard c where u.id_no_hash=(select card_no_hash from `order` where card_no_hash=?) and u.mobile_hash=? and u.id=1

LogicSQL:
SELECT u.mobile, u.username, u.id_no FROM user `u`, bankcard c where u.id_no in (select card_no from `order` where card_no=?) and u.mobile=? and u.id=1

ActualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no FROM user `u`, bankcard c where u.id_no_hash in (select card_no_hash from `order` where card_no_hash=?) and u.mobile_hash=? and u.id=1

LogicSQL:
SELECT u.mobile, u.username, u.id_no FROM user `u`, bankcard c where (select card_no from `order` where card_no=?)=u.id_no and u.mobile=? and u.id=1

ActualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no FROM user `u`, bankcard c where (select card_no_hash from `order` where card_no_hash=?)=u.id_no_hash and u.mobile_hash=? and u.id=1

LogicSQL:
SELECT u.mobile, u.username, u.id_no FROM user `u`, bankcard c where (select card_no from `order` where card_no=?)=(select id_no from `user` where id_no=?) and u.mobile=? and u.id=1

ActualSQL:
SELECT u.mobile_cipher AS mobile, u.username, u.id_no_cipher AS id_no FROM user `u`, bankcard c where (select card_no_hash from `order` where card_no_hash=?)=(select id_no_hash from `user` where id_no_hash=?) and u.mobile_hash=? and u.id=1

### PR DESCRIPTION
**Pull request by cheese8**
Fixes #12825

Changes proposed in this pull request:

Support not nested subquery (from SetAssignment and Expression) on UpdateStatement

### CODE REVIEW

**Comment by cheese8** [comment]
@strongduanmu Could you please help on review it?

**Comment by strongduanmu** [comment]
> @strongduanmu Could you please help on review it?

@cheese8 Good job, I will review this pr.

File: shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/encrypt/case/update_for_query_with_cipher.xml [hunk 1]
```diff
@@ -36,4 +36,14 @@
...
         <output sql="UPDATE t_account_bak SET account_id = 1, cipher_certificate_number = 'encrypt_111X', assisted_query_certificate_number = 'assisted_query_111X', plain_certificate_number = '111X', cipher_password = 'encrypt_aaa', assisted_query_password = 'assisted_query_aaa', plain_password = 'aaa', cipher_amount = 'encrypt_1000', plain_amount = 1000, status = 'OK' WHERE account_id = 1 AND assisted_query_certificate_number = 'assisted_query_111X' AND assisted_query_password = 'assisted_query_aaa' AND cipher_amount = 'encrypt_1000' AND status = 'OK'" />
     </rewrite-assertion>
+
+    <rewrite-assertion id="update_subquery_projections" db-types="MySQL">
+        <input sql="UPDATE t_account_bak SET account_id = 1, certificate_number = (select certificate_number from t_account where certificate_number='111X'), password = 'aaa', amount = 1000, status = 'OK' WHERE account_id = 1 AND certificate_number = '111X' AND password = 'aaa' AND amount = 1000 AND status = 'OK'" />
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/encrypt/case/update_for_query_with_cipher.xml:41]: @cheese8 This PR does not seem to be able to perfectly support the subqueries contained in the update statement. In this case, the rewritten sql should report an error when executed on the database.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by strongduanmu** [comment]
@cheese8 Since the subqueries in the update statement cannot be perfectly supported for the time being, I will close this pr, and we will consider using the federation execution engine to support such complex statements.

**Pull request closed without merging**